### DISCUSSION TITLE
FEATURE REQUEST: Register Model Observers via Model property.

### DISCUSSION THREAD

**Discussion by trip-somers** [laravel/framework#34502] [General]
Typically, model observers are registered in a service provider or in the model itself via an overloaded `boot()` method that looks like this:

```php
    protected static function boot()
    {
        parent::boot();

        self::observe(new DummyObserver);
    }
```

This request seems pretty obvious, so there's probably something that I've missed, but could the base `boot()` method be updated to call a `bootObservers()` method -- similar to the `bootTraits()` method -- that will process an `$observers` array property so that the above can be avoided and replaced with something that looks like this:

```php
    protected $observers = [DummyObserver::class];
```

(I made it an array and plural since I can't recall if there's support or need for multiple observers on a single model. It can easily be singular if that makes the most sense.)

**Comment by browner12** [comment]
I believe this is possible already.

https://github.com/laravel/framework/blob/8.x/src/Illuminate/Database/Eloquent/Concerns/HasEvents.php

**Reply by trip-somers** [reply]
`$observables` appears to be for extending the standard model events, not registering observers.

If I am incorrect, would you please share a brief example?

**Comment by ecrmnn** [comment]
Bump.

Would definitely like to have this in the framework without having to use `boot` or `booted`.

Registering observers in the EventServiceProvider makes it harder to debug.

**Comment by pascalvgemert** [comment]
I created a trait called `Observable` which can be used on your models.

**Example:**

```php
namespace App\Models\Traits;

use RuntimeException;

/**
 * @mixin \Illuminate\Database\Eloquent\Model
 */
trait Observable
{
    public static function bootObservable(): void
    {
        $baseName = class_basename(static::class);
        $observableClass = "\\App\\Observers\\{$baseName}Observer";

        if (! class_exists($observableClass)) {
            throw new RuntimeException("Observer class {$observableClass} does not exist.");
        }

        (new static())->registerObserver($observableClass);
    }
}
```

I will try to create a package for it somewhere next week.

**Comment by henzeb** [comment]
It's now possible to do this with PHP Attributes. https://laravel.com/docs/10.x/eloquent#observers
